doc/adr/0007-marconi-indexer.rst

```diff
@@ -0,0 +1,118 @@
+
+We make a distinction between the volatile blocks which are stored in memory as events (and are derived from blocks). We fold these events into the aggregated on-disk data structure for which we no do not require to keep multiple versions (rollbacks cannot happen for this data structure).
+
+We improve on that idea by allowing part of the volatile blocks to be stored on disk. While this is not required at the API level, the usage pattern would be to have a set of events, as well as the aggregated data structure stored on disk. The compromise here is that the more data is on the disk, the more we will need to work with the disk and the slower the indexing process will become. The advantage would be reduced memory usage.
+
+Events
+^^^^^^
+
+Events need to contain information about the slot number and block id when they were produced.
+
+The slot number information is used in case of rollbacks (when we only get the old slot number that we need to rollback to) and for resuming the operation of the index, in which case we need both the slot numbers and block ids.
+
+Note that to support resume from disk we need to always have at least one event persisted on disk which contains the slot number and block id from which we are supposed to resume operation. In case there are more than one events stored on disk we can use those as resume alternatives in the case of the ChainSync protocol.
+
+Queries
+^^^^^^^
+
+We also extend the queries with a bit more structure that will make specifying query validity intervals possible. The validity is important in the case where we want to query several indexers and we would like them to require to have processed all the information up to some slot number or be between some slot interval.
+
+At this point all query results are synchronous. We have plans to extend this functionality, but these plans are based on updating the notification system for indexers which will be described in a further ADR.
+
+API Design
+^^^^^^^^^^
+
+We want the API for our users to be as flexible as possible, so some of our previously mentioned design patterns are not captured by the API, but rather by its implementation.
+
+Data types
+""""""""""
+
+* Events
+  ** Slot numbers (requires Ord)
+  ** Block id (requires Eq)
+  ** e (type variable standing for the event)
+* Query
+  ** Validity interval (can be always valid)
+  ** q (type variable standing for the query)
+* Result
+  ** Slot number at which the query was ran
+  ** r (type variable standing for the query result)
+
+Functions
+"""""""""
+
+* Query
+  ** Indexer
+  ** Validity interval
+  ** The query (q type variable)
+  ** Returns the result
+* Store
+  ** Indexer
+  ** Does not return anything useful
+* Resume
+  ** Indexer
+  ** Returns a list of slot numbers and block ids
+
+Runtime parameters
+""""""""""""""""""
+
+* Minimum events retained (this should be the previously mentioned K parameter)
+* Maximum in-memory events (should be less than K)
+
+Extension mechanisms
```

Comment:
What's the point of this section? How is it related to the decision?

Reply:
We were forced into the decision because our implementation was difficult to customise, so I think it is worth the bytes to explain what this new design allows us to customise.